[nT2] rechits reconstruction from digis

forthommel opened the pull request:
#### PR description:

In this PR we introduce the Totem (for now, new T2-only) geometry-DetId mapping condition object, and the construction of a T2 rechits collection from the digis collection recently introduced in #38886.
In order to avoid yet another duplication of pre-existing digi-rechits conversion code, the algorithmic part of this latter was transformed into a `TimingRecHitProducerAlgorithm` base class from which the pre-existing `CTPPSDiamondRecHitProducerAlgorithm`, `TotemTimingRecHitProducerAlgorithm`, and the newly-introduced `TotemT2RecHitProducerAlgorithm` inheritate.
Most of the development history has been stashed in a few minimally-working commits, and follow-up developments were rebased on top of this branch.

#### PR validation:

Code compiles and runs on top of a digi-level sample. Matrix tests succede.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for: N/A

forthommel committed:
DF+Reco: Added rechits skeletton and dictionary generation part for rechits-derived objects

forthommel committed:
DF+Reco: Added a few bits to the reco algorithm logic, TotemT2RecHitProducerAlgorithm now inherits from timing base one

forthommel committed:
Geom: Added an ESProducer for the TotemGeometry intermediate object

forthommel committed:
General: added example reconstruction configuration

forthommel committed:
Geom: Totem geometry has its own record

forthommel committed:
Geom: ESModule now accepts both DDL and DD4hep

forthommel committed:
Geom: Populating the geometry elements substructure from the DDL/DD4hep hierarchy

forthommel committed:
DF: Updated detid, rechits, and steering the rechits producer algorithm
to produce the output collection

forthommel committed:
Geom: Dropping DD4hep dependency and using ROOT geometry definition
instead

forthommel committed:
DF: Added ToT for T2 rechits

forthommel committed:
General: Added time calibration to rechits builder

forthommel committed:
Reco: timing calibration LUT also loaded for T2 RH producer (for consistency)

forthommel committed:
Reco: added rechits-only production test

forthommel commented on DataFormats/TotemReco/src/classes_def.xml:
```diff
@@ -7,4 +7,12 @@
   <class name="std::vector<TotemT2Digi>"/>
   <class name="std::vector<edm::DetSet<TotemT2Digi> >"/>
+  <class name="TotemT2RecHit" ClassVersion="2">
+    <version ClassVersion="2" checksum="969691996"/>
```
Class version may be bumped to 10 in a follow-up push? Not sure what is the current convention nowadays.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38986/31440

- This PR adds an extra 76KB to repository

cmsbuild commented:
A new Pull Request was created by  @forthommel (Laurent Forthomme) for master.

It involves the following packages:

- DataFormats/TotemReco (**reconstruction**)
- Geometry/ForwardGeometry (**geometry**)
- Geometry/Records (**geometry**)
- RecoPPS/Local (**reconstruction**)


@civanch, @Dr15Jones, @makortel, @ianna, @clacaputo, @cmsbuild, @bsunanda, @mdhildreth, @jpata can you please review it and eventually sign? Thanks.
@bsunanda, @rovere, @fabiocos, @grzanka this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26688/summary.html
**COMMIT**: a4dc9da37399593fc432403a013473a1489809c4
**CMSSW**: CMSSW_12_5_X_2022-08-07-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38986/26688/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3691612
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3691582
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-07-2300+5000cf/52013/triggerResults): no differences found

clacaputo commented:
Hi @forthommel , is this PR planned to be integrated into 12_4_X?

forthommel commented:
Hi @clacaputo ! No, it will only used under specific pp data-taking conditions in ~october-november, and for HI runs later on. So, 12_5_X is all fine.

makortel commented on Geometry/Records/interface/TotemGeometryRcd.h:
```diff
@@ -0,0 +1,28 @@
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/Records/interface/PTotemRcd.h"
+
+#include "CondFormats/AlignmentRecord/interface/RPRealAlignmentRecord.h"
```
This `#include` appears to be unused.

forthommel committed:
Geom+DF+Reco: several adjustments requested from reviewers

- replaced edm::DetSetVector to edmNew::DetSetVector for T2 rechits production
- zero-initialisation of rechits members
- subsequently TimingRecHitProducerAlgorithm object can now take any
  input/output type
- downgraded TotemT2Digi and TotemT2RecHit versions to 3
- timing rechits production algorithm now stores a pointer to
  calibrations instead of a copy

Additionally:

- TotemTimingRecHitProducer now also derives from the
  newly-introduced TimingRecHitProducerAlgorithm base object
- making better use of improved accessors-from-token for
  collections in event for the three amended rechits producers.
  This allows a better code readibility generally.

clacaputo commented on DataFormats/TotemReco/src/classes_def.xml:
```diff
@@ -7,4 +7,12 @@
   <class name="std::vector<TotemT2Digi>"/>
   <class name="std::vector<edm::DetSet<TotemT2Digi> >"/>
+  <class name="TotemT2RecHit" ClassVersion="2">
+    <version ClassVersion="2" checksum="969691996"/>
```
Hi @makortel , for my education, where I can find more info on this?

forthommel replied:
@makortel thanks for your review! Indeed, I did not find much info about this myself, and mostly copied other data formats conventions. Is the `TotemT2Digi` version 10 above fine too, or should it be decremented?

jpata replied:
I stumbled on this as well a few times. https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning

> Version numbers 0, 1, and 2 have special meaning to ROOT and should not be used as version numbers. By convention, CMS will use 3 as the initial version number of a class which has never been stored before. 

Something that could as well be caught via a test...

forthommel replied:
I believe I can also decrement the `TotemT2Digi` version to 3 then!

forthommel commented on RecoPPS/Local/interface/TimingRecHitProducerAlgorithm.h:
```diff
@@ -0,0 +1,46 @@
+  virtual ~TimingRecHitProducerAlgorithm() = default;
+
+  void setCalibration(const PPSTimingCalibration& calib, const PPSTimingCalibrationLUT& calibLUT) {
+    calib_ = calib;
+    calibLUT_ = calibLUT;
```
Good point. Although this method was used all along run 2 for the PPS diamond reconstruction scheme. Surely this can be optimised a bit (even though the objects are not that large, and the number of channels remain limited...)

forthommel replied:
Now I realise I forgot to address this in my last push. Will come with a follow-up in a short while. In the meantime the latest additions can already be tested.

forthommel replied:
Done with 2e0936e8bef.

forthommel commented on Geometry/ForwardGeometry/src/TotemGeometry.cc:
```diff
@@ -0,0 +1,65 @@
+  if (in_t2)
+    browseT2(parent);
+  // start the recursive browsing
+  for (const auto& child : parent->components())
+    browse(const_cast<const DetGeomDesc*&>(child), in_t2);
```
Not necessarily. In a follow-up push I adapted the `TotemGeometry::browse` method to accept a const pointer instead of a const reference to a pointer, and thus the recursive browsing is simplified.

forthommel commented on Geometry/ForwardGeometry/src/TotemGeometry.cc:
```diff
@@ -0,0 +1,65 @@
+#include "Geometry/ForwardGeometry/interface/TotemGeometry.h"
+
+TotemGeometry::TotemGeometry(const DetGeomDesc* dgd) {
+  std::deque<const DetGeomDesc*> buffer;
+  buffer.emplace_back(dgd);
+  while (!buffer.empty()) {
+    // get the next item
+    const auto* desc = buffer.front();
+    buffer.pop_front();
+    browse(desc, false);
+  }
```
I wanted to follow a similar convention to the PPS geometry definition (as the `TotemGeometry` object payload may contain more subdetectors in the future), but true that your proposal simplifies it greatly for the time being.

makortel commented on DataFormats/TotemReco/src/classes_def.xml:
```diff
@@ -7,4 +7,12 @@
+    <version ClassVersion="2" checksum="969691996"/>
+  </class>
+  <class name="edm::DetSet<TotemT2RecHit>"/>
+  <class name="edm::DetSetVector<TotemT2RecHit>"/>
```
> 1\. would the API change for all subsequent consumer code using this format?

It likely would, depending on how the `edm::DetSetVector` is being used. IIRC just iterating over the elements is not much different, but some other access patterns could be more different (or not available).

> 2\. would a migration campaign from `edm::DetSetVector<T>` to `edmNew::DetSetVector<T>` be necessary for all PPS objects in `DataFormats/CTPPSReco` and `DataFormats/CTPPSDigi`? 

If such a change would be made, yes, it would be good to replace all uses of `edm::DetSetVector` with `edmNew::DetSetVector` (and that would certainly be beyond this PR). Unfortunately it would also break backwards compatibility, which means such the moment for such a change would need some planning.

> Alternatively, is there any plan to "simply" replace `edm::DetSetVector<T>` by its `edmNew::DetSetVector<T>` (with all the conversion codes it implies)?

We have long hoped the `edm::DetSetVector` could eventually be phased out (#22711), but have not taken any real action towards that (lower hanging fruits, people usually get annoyed by data format backwards incompatibilities). But `edmNew::DetSetVector` is not (and can't really be made) as a drop-in replacement of `edm::DetSetVector` (not sure if you meant this, but just to make sure).

forthommel replied:
Indeed, I can tell these access patterns are much different (and would require an overall improvement in the API to be made convenient IMHO). In order not to add a few unnecessary signatures in the heat of the Summer holidays I will also replace the `edm::DetSetVector<TotemT2Digi>` collection by a `edmNew::DetSetVector<TotemT2Digi>` in a follow-up PR.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38986/31512

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #38986 was updated. @civanch, @Dr15Jones, @makortel, @ianna, @clacaputo, @cmsbuild, @bsunanda, @mdhildreth, @jpata can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38986/31513

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #38986 was updated. @civanch, @Dr15Jones, @makortel, @ianna, @clacaputo, @cmsbuild, @bsunanda, @mdhildreth, @jpata can you please check and sign again.

forthommel commented:
Ready for another batch of reviews and/or a `please test`.

clacaputo commented:
@cmsbuild please test

forthommel commented:
Thanks, @clacaputo ! Although the tests can now be aborted, I found an obvious problem with the new `TimingRecHitsProducerAlgorithm` which is about to be solved in a follow-up push.

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/summary.html
**COMMIT**: 2e0936e8befe0652fbd391719826fb83be2b3946
**CMSSW**: CMSSW_12_5_X_2022-08-10-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38986/26753/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [138.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/138.4_PromptCollisions+RunMinimumBias2021+ALCARECOPROMPTR3+HARVESTDPROMPTR3)
```
----- Begin Fatal Exception 11-Aug-2022 12:31:16 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
- [138.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3)
```
----- Begin Fatal Exception 11-Aug-2022 12:32:55 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
- [136.731](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2_skimSinglePh)
```
----- Begin Fatal Exception 11-Aug-2022 12:33:17 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [136.793](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/136.793_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017_skimDoubleEG)
- [136.874](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/136.874_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM)
- [140.56](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18)
- [139.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/139.001_RunMinimumBias2021+RunMinimumBias2021+HLTDR3_2021+RECODR3_MinBiasOffline+HARVESTD2021MB)
- [1306.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15)
- [312.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/312.0_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO)
- [1330.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15)
- [10042.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/10042.0_ZMM_13+2017+ZMM_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [10024.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/10024.0_TTbar_13+2017+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [25202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25)
- [10824.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/10824.0_TTbar_13+2018+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [11634.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/11634.7_TTbar_14TeV+2021_trackingMkFit+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano)
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/11634.0_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [11634.914](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/11634.914_TTbar_14TeV+2021_DDDDB+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [10224.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/10224.0_TTbar_13+2017PU+TTbar_13TeV_TuneCUETP8M1_GenSim+DigiPU+RecoFakeHLTPU+HARVESTFakeHLTPU+Nano)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [11634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/11634.911_TTbar_14TeV+2021_DD4hep+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA)
- [39500.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/39500.0_CloseByPGun_CE_H_Coarse_Scint+2026D88+CE_H_Coarse_Scint_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39496.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/39496.0_CloseByPGun_CE_E_Front_120um+2026D88+CE_E_Front_120um_GenSimHLBeamSpotHGCALCloseBy+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [11834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/11834.0_TTbar_14TeV+2021PU+TTbar_14TeV_TuneCP5_GenSim+DigiPU+RecoNanoPU+HARVESTNanoPU)
- [23234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39434.75](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/39434.75_TTbar_14TeV+2026D88_HLT75e33+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HLT75e33+HARVESTGlobal)
- [35034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/35034.0_TTbar_14TeV+2026D77+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/39434.0_TTbar_14TeV+2026D88+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [28234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [250202.181](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25)
- [39634.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrix-results/39634.999_TTbar_14TeV+2026D88PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU)
</details>



## RelVals-INPUT

- [138.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/138.4_PromptCollisions+RunMinimumBias2021+ALCARECOPROMPTR3+HARVESTDPROMPTR3)```138.4_PromptCollisions+RunMinimumBias2021+ALCARECOPROMPTR3+HARVESTDPROMPTR3/step2_PromptCollisions+RunMinimumBias2021+ALCARECOPROMPTR3+HARVESTDPROMPTR3.log```
- [136.9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/136.9_RunDoubleMuon2016C+RunDoubleMuon2016C+RAWRECOTE16+RAWRECOLHECLEANTE16+EMBEDHADTE16+EMBEDMINIAOD16)```136.9_RunDoubleMuon2016C+RunDoubleMuon2016C+RAWRECOTE16+RAWRECOLHECLEANTE16+EMBEDHADTE16+EMBEDMINIAOD16/step2_RunDoubleMuon2016C+RunDoubleMuon2016C+RAWRECOTE16+RAWRECOLHECLEANTE16+EMBEDHADTE16+EMBEDMINIAOD16.log```
- [138.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3)```138.5_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3/step2_ExpressCollisions+RunMinimumBias2021+TIER0EXPRUN3+ALCARECOEXPR3+HARVESTDEXPR3.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [140.56](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18)
- [140.57](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/140.57_RunHI2018Reduced+RunHI2018Reduced+RECOHID18+HARVESTDHI18)
- [1041.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1041.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCAL+ALCASPLITPPSCAL+ALCAHARVDPPSCAL)
- [1043.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1043.0_RunRawPPS2022B+RunRawPPS2022B+TIER0EXPPPSCALRUN3+ALCASPLITPPSCALSAMPIC+ALCAHARVDPPSCALSAMPIC)
- [1030.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1030.0_RunHLTPhy2017B+RunHLTPhy2017B+TIER0EXPHPBS+ALCASPLITHPBS+ALCAHARVDHPBS+ALCAHARVDHPBSLOWPU)
- [1044.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1044.0_RunRawPPS2022B+RunRawPPS2022B+TIER0EXPPPSCALRUN3+ALCASPLITPPSCALRUN3+ALCAHARVDPPSCALRUN3)
- [1042.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1042.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCALALIG+ALCASPLITPPSALIG+ALCAHARVDPPSALIG)
- [1001.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1001.3_RunSingleMuon2022B+RunSingleMuon2022B+TIER0EXPRUN3+ALCAEXPRUN3+ALCAHARVDEXPRUN3)
- [1001.2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1001.2_RunZeroBias2017F+RunZeroBias2017F+TIER0EXPRUN2+ALCAEXPRUN2+ALCAHARVDSIPIXELCAL+ALCAHARVDSIPIXELCALLA+ALCAHARVD4+ALCAHARVDSIPIXELALIHG)
- [1002.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1002.3_RunZeroBias2022B+RunZeroBias2022B+TIER0PROMPTRUN3+ALCASPLITRUN3+ALCAHARVDEXPRUN3)
- [1040.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/runTheMatrixINPUT-results/1040.0_RunZeroBias2017F+RunZeroBias2017F+TIER0RAWSIPIXELCAL+ALCASPLITSIPIXELCAL+ALCAHARVDSIPIXELCAL)
</details>



## AddOn Tests

- [hlt_mc_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_mc_Fake2)
```
----- Begin Fatal Exception 11-Aug-2022 12:34:57 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_mc_GRun)
```
----- Begin Fatal Exception 11-Aug-2022 12:36:58 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_mc_HIon)
```
----- Begin Fatal Exception 11-Aug-2022 12:35:57 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
MissingParameter: Parameter 'timeSliceNs' not found.
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more addon errors ...</summary>

- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_mc_PIon)
- [hlt_mc_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_mc_PRef)
- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_data_Fake2)
- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_data_GRun)
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_data_HIon)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_data_PIon)
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26753/addOnTests/hlt_data_PRef)
</details>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38986/31517

- This PR adds an extra 96KB to repository

cmsbuild commented:
Pull request #38986 was updated. @civanch, @Dr15Jones, @makortel, @ianna, @clacaputo, @cmsbuild, @bsunanda, @mdhildreth, @jpata can you please check and sign again.

clacaputo commented:
@cmsbuild please test

forthommel commented:
@cms-sw/reconstruction-l2, @cms-sw/geometry-l2, any plan to tackle this PR in the forthcoming days?

jpata commented:
Vacations affect our review turnaround time during summer, but I think it's on track for the next release.

- Which workflows does the new code run in?
- Are any reco changes expected?
- What is the CPU/mem profiling performance of the new code?

forthommel commented:
Many thanks for your fast reaction, @jpata! And indeed I know the Summer can certainly affect the various workflows, hence my kind "ping" :)
No workflow currently contain this added reconstruction scheme, as the detector is not yet integrated into cDAQ and we can only rely on simulated frames. Consequently, no reco changes are expected for standard WFs, and CPU/memory profiling is relatively difficult to evaluate outside thumbrule/scaling (only 64 readout channels are to be unpacked for low-PU end-of-year conditions).
However, we can surely provide a recipe to steer the newly-introduced rechits producer algorithm, and we are more than open to discuss the rechits data format if necessary.

jpata commented on DataFormats/TotemReco/interface/TotemT2RecHit.h:
```diff
@@ -0,0 +1,40 @@
+
+private:
+  /// Tile centre position
+  GlobalPoint centre_;
+  /// Leading edge time
+  float time_;
+  /// Uncertainty on leading edge time
+  float time_unc_;
+  /// Time over threshold/pulse width
+  float tot_;
```
should these be zero-initialized?

jpata commented:
> No workflow currently contain this added reconstruction scheme, as the detector is not yet integrated into cDAQ and we can only rely on simulated frames.

It might be useful to consider some kind of regular testing (e.g. a sim-only workflow) for this.

Can you assess the impact on the data size?

civanch commented:
@forthommel , you will also nee to squash commits at the end.

forthommel commented:
@civanch I would suggest to edit [the FAQ](http://cms-sw.github.io/faq.html#how-do-i-collapse-multiple-commits-into-one) to match with the current release managers mindset where the squashing, effectively destroying the development history of a full project, is enforced. For this particular feature development, we already collapsed 50+ commits into about 10 self-contained ones before review, but it seems that your strategy remains a mystery (at least for me).

civanch commented:
@forthommel , I did not wrote FAQ and would not change there anything myself. What is important - initialize class members at construction as @jpata suggested.

I also guess nT2 data is relatively small, even if all scintillators will be fired.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38986/31644

- This PR adds an extra 88KB to repository

cmsbuild commented:
Pull request #38986 was updated. @civanch, @Dr15Jones, @makortel, @ianna, @clacaputo, @cmsbuild, @bsunanda, @mdhildreth, @jpata, @mandrenguyen can you please check and sign again.

forthommel commented:
Thanks! I just zero-initialised the rechits members in the last push.

For the rate, based on the fact the nT2 unpacking under low-intensity (PU << 1), each BX will deliver at most 8 (planes) * 4 (channels) * 2 (arms) = 64 rechits, with 3 (`GlobalPoint`) + 3 (time, time uncertainty, tot) = 6 floats/rechit.
Hence the maximal raw (uncompressed, prior to L1/HLT) is 64 * 6 * 4 (bytes/float on 64-bit) = 1536 bytes/BX = 1.5 kB/BX.

jpata commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5000cf/26922/summary.html
**COMMIT**: 65b6a5e64a3c12485440d682d129693f2f863feb
**CMSSW**: CMSSW_12_5_X_2022-08-18-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38986/26922/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3693040
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3693016
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-18-2300+5000cf/52217/triggerResults): no differences found

forthommel commented:
> > No workflow currently contain this added reconstruction scheme, as the detector is not yet integrated into cDAQ and we can only rely on simulated frames.
> 
> It might be useful to consider some kind of regular testing (e.g. a sim-only workflow) for this.
> 
> Can you assess the impact on the data size?

Hi @jpata, I hope I managed to clarify our use case with https://github.com/cms-sw/cmssw/pull/38986#issuecomment-1220471407. We will work on a more automated testing procedure as soon as our first emulated packed frames will be delivered.

forthommel commented:
Just a kindly reminder. Are we all set for a new batch of reviews?

jpata commented:
>  GlobalPoint centre_;

did you check that this class is zero-initialized by the default constructor?

forthommel commented:
Following the `typeset`s spaghetti of this common geometry object, [a comment in `Point3DBase`](https://github.com/cms-sw/cmssw/blob/master/DataFormats/GeometryVector/interface/Point3DBase.h#L19-L22) pretends it is for builtin types, although it depends if `USE_SSEVECT` is defined in the standard build environment as [a comment in SSE's `Basic3DVector` default constructor](https://github.com/cms-sw/cmssw/blob/master/DataFormats/GeometryVector/interface/private/sseBasic3DVector.h#L39-L42) is slightly less assertive...

jpata commented:
+reconstruction
- work towards Totem reconstruction
- does not currently run in any workflow or test

forthommel commented:
Thanks, @jpata ! Now waiting for geometry reviewers' +1. @Dr15Jones, @civanch, @bsunanda, @ianna, @makortel, @mdhildreth?

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

forthommel commented on DataFormats/TotemReco/src/classes_def.xml:
```diff
@@ -1,10 +1,18 @@
 <lcgdict>
-  <class name="TotemT2Digi" ClassVersion="10">
-    <version ClassVersion="10" checksum="4024006040"/>
+  <class name="TotemT2Digi" ClassVersion="3">
```
Given no single dataset was created with this extra `TotemT2Digi` object, I thought it would be a trivial change.

perrotta replied:
> Given no single dataset was created with this extra `TotemT2Digi` object, I thought it would be a trivial change.

Yes, quite probably you are right. But let me wait for an answer from the experts of the framework

makortel replied:
If you are sure that there are no data with the class version 10 (that should ever be read after this change), changing the class version is acceptable (even if the change would otherwise look suspicious). The value `3` leads to a little bit smaller memory usage than `10`, so the change has some positive impact.

forthommel commented on RecoPPS/Local/src/TotemT2RecHitProducerAlgorithm.cc:
```diff
@@ -0,0 +1,52 @@
+*
+****************************************************************************/
+
+#include "FWCore/Utilities/interface/isFinite.h"
```
See https://github.com/forthommel/cmssw/blob/nt2_rechits_geom-12_5_X/RecoPPS/Local/src/TotemT2RecHitProducerAlgorithm.cc#L40.

forthommel commented on RecoPPS/Local/plugins/TotemT2RecHitProducer.cc:
```diff
@@ -0,0 +1,98 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
The `ESHandle`, or all 4 includes?

perrotta replied:
> The `ESHandle`, or all 4 includes?

 `ESHandle

forthommel commented on Geometry/ForwardGeometry/src/TotemGeometry.cc:
```diff
@@ -0,0 +1,56 @@
+      tile += 1;
+    else
+      plane += 1;
+    tile = tile / 2 - 1;
```
This case is not encountered in @bsunanda's [T2 geometry definition](https://github.com/cms-sw/cmssw/pull/29268). May we introduce this extra check in a follow-up PR?

perrotta replied:
Are you claiming that `tile` cannot be 0 or 1 by construction? Where is it coded? 
If so, either a check or, at least, a line of comment could be useful

forthommel replied:
By construction, indeed. Fine with me to introduce this extra check, but we would appreciate a fast review on a follow-up push, as `12_5_X` will be used for data taking and the DQM PR is ready to be submitted.

perrotta commented on Geometry/ForwardGeometry/src/TotemGeometry.cc:
```diff
@@ -0,0 +1,56 @@
+  else if (parent->name() == "TotemT2Scint") {
+    unsigned short plane = 2 * (parent->copyno() / 10);
+    unsigned short tile = parent->copyno() % 10;
+    if (tile % 2 != 0)
+      tile += 1;
+    else
+      plane += 1;
+    tile = tile / 2 - 1;
```
```suggestion
    if (tile % 2 == 0)
      plane += 1;
    tile = (tile+1) / 2 - 1;
```
looks more intuitive to me, and easier to understand that once `tile = 0` is excluded the expression remains correct.

(In any case, unless there is a suggestion to restore the previous ClassVersion here above, I agree to move all these fixes in a bugfix PR that can be merged even after pre5)

forthommel replied:
Many thanks, @perrotta ! Indeed, if we could avoid another round of `+1`s in this holiday period, things would be easier :)
